complex maintenance stuff + choleski

lichtkind opened the pull request:
docs tests, streamlining

pierre-vigier commented:
Test are not passing, i'll wait a bit to merge

lichtkind commented:
sure i pleaing with somthing i never tried before

lichtkind commented:
preparing a .table method which is s sort of .perl but with a table view like in octave

lichtkind commented:
choleski is commented out because  i need another method to test matrices if they are positive, or there is no choleski decomp

pierre-vigier commented:
Merging that part first,

thanks!

pierre-vigier commented on lib/Math/Matrix.pm6:
```diff
@@ -181,6 +188,18 @@ method is-orthogonal(Math::Matrix:D: --> Bool) {
 
+
+method is-positive-definite (Math::Matrix:D: --> Bool) { # with Sylvester's criterion
+    return False unless self.is-square or min self.size < 1;
```
why min self.size < 1 ?
Normal we should never have a size = 0 , neither line nor column, i will be at least 1

lichtkind replied:
allright i just wanted for security but 0 *0 matrices are thinkable

lichtkind replied:
i take that out with next change.

pierre-vigier commented on lib/Math/Matrix.pm6:
```diff
@@ -10,9 +11,9 @@ method !column-count { $!column-count }
-    die "All Row must contains only numeric values" unless all( @m[*;*] ) ~~ Numeric;
+    fail "Expect an Array of Array" unless all @m ~~ Array;
+    fail "All Row must contains the same number of elements" unless @m[0] == all @m[*];
+    fail "All Row must contains only numeric values" unless all( @m[*;*] ) ~~ Numeric;
```
Fail will not be fatal, which means, that if not check, the execution can continue, are you sure we want that?

lichtkind replied:
actually thats what i prefer or at least throw a handable eqception , i tought this would die unless i handle it

lichtkind replied:
i take it out for now but i dont like routines that can take out the whole app without any way to catch error.

pierre-vigier commented on lib/Math/Matrix.pm6:
```diff
@@ -181,6 +188,18 @@ method is-orthogonal(Math::Matrix:D: --> Bool) {
+method is-positive-definite (Math::Matrix:D: --> Bool) { # with Sylvester's criterion
+    return False unless self.is-square or min self.size < 1;
+    return False unless self.determinant > 0;
+    my $sub = self.cone;
```
self.clone ?
If yes, what would it do?

lichtkind replied:
gives me matrix object with same values. its not perfect solution but I don't know how else handle logic and references.

pierre-vigier commented on lib/Math/Matrix.pm6:
```diff
@@ -51,6 +52,15 @@ method zero(Math::Matrix:U: Positive_Int $rows, Positive_Int $cols = $rows) {
+    @clone.splice($row,1);
+    @clone = map { $^r.splice($col, 1); $^r }, @clone;
+    Math::Matrix.new( @clone );
+}
```
not sure to understand that one, let take matrix a, like following:

1|2|3|4|5
---|---|---|---|---
6|7|8|9|10
11|12|13|14|15
16|17|18|19|20

a.submatrix( 1,2) should be a without row 1 and column 2? like that?:

1|2|4|5
---|---|---|---|---
11|12|14|15
16|17|19|20

Am i understanding right?
it would correspond to that definition: https://proofwiki.org/wiki/Definition:Submatrix

However, i was planning to create a method submatrix( begin_row, begin_column, end_row, end_column ), 
with the same starting matrix a, we would have, a.submatrix(1,1,3,5) which would keep everything between cell 1,1 and 3,5, that is:

7|8|9|10
---|---|---|---
12|13|14|15
17|18|19|20
 
That definition is also a possible one apparently, from for instance there: http://mathworld.wolfram.com/Submatrix.html

Need some clarification there, which one should we implement? both? with what name? as the name is referred for both definition apparently

lichtkind replied:
my problem is that we in german use completely different names and i did quick lookup. i fully support your proposed function but would call it differently.

lichtkind commented:
ran into sack of problem becasue vclone is not deep, i will have to make my own clone

lichtkind commented:
i have to make own clone first to fix methods who already use it,
i propose to call it AoA clone since it only works on that structure

lichtkind commented:
sure i can some things take out but the cloning stuff has to go in because it actually fixes broken stuff

yes i added som minor things but also fixed typos just tell me what to remove

actually even in german the proper term seems to be submatrix mean in the sense i used it. but i have to check more sources.

lichtkind commented:
waite a sec there is one last typo
